core/src/main/java/org/apache/iceberg/io/RollingFileWriter.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A rolling writer capable of splitting incoming data/deletes into multiple files within one spec/partition.
+ */
+abstract class RollingFileWriter<T, W extends FileWriter<T, R>, R> implements FileWriter<T, R> {
+  private static final int ROWS_DIVISOR = 1000;
+
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final FileFormat fileFormat;
+  private final long targetFileSizeInBytes;
+  private final PartitionSpec spec;
+  private final StructLike partition;
+
+  private EncryptedOutputFile currentFile = null;
+  private W currentWriter = null;
+  private long currentRows = 0;
+
+  private boolean closed = false;
+
+  protected RollingFileWriter(OutputFileFactory fileFactory, FileIO io, FileFormat fileFormat,
+                              long targetFileSizeInBytes, PartitionSpec spec, StructLike partition) {
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.fileFormat = fileFormat;
+    this.targetFileSizeInBytes = targetFileSizeInBytes;
+    this.spec = spec;
+    this.partition = partition;
+  }
+
+  protected abstract W newWriter(EncryptedOutputFile file);
+
+  protected abstract long length(W writer);
```

Comment:
can we add a `long length()` API to to  the `FileWriter` interface and remove this method?

Reply:
I wanted to but we have `SortedPosDeleteWriter` that does not support giving the length. Maybe, we should just add `length` to `FileWriter` and throw an exception whenever it is not supported. Thoughts, @stevenzwu?

Reply:
yeah. I am in favor of having `SortedPosDeleteWriter` throwing not supported exception